### ISSUE TITLE
Feedback - no way to wait for subtasks to complete

### ISSUE THREAD

**Issue by tom-leys** [nmcclatchey/ThreadPool#6]
Thanks for creating this pool and publishing it for the world! I really appreciate it.

I was considering using it as a replacement for our previous ThreadPool (which does not do work stealing and is very naive). Our current pool returns a `std::future<return_type>` when a task is enqueued for later execution. 

We tend to collect the futures into a list and then wait on them from the spawning thread. I was hoping that you would provide a pool which allowed the following

1. Outer thread creates a number of tasks, some of which might run immediately in the outer thread if the pool is super busy
2. Outer thread eventually reaches and enters some sort of "wait on tasks" routine which waits on one or more subtasks to complete
3. Both threadpool and outer thread stay busy working on subtasks until the wait is done (and the outer thread has also completed its current task)

I think that I'm describing something like the "blocking style" from [TBB](https://software.intel.com/content/www/us/en/develop/documentation/tbb-documentation/top/intel-threading-building-blocks-developer-reference/task-scheduler.html )

I couldn't see how the outer thread is supposed to wait on tasks in the pool to complete (note I would prefer you could wait on a subset of the pool, not the whole pool).

Perhaps I am mis-understanding work stealing, for instance does a call to `pool.schedule()` not "return" until the given task is done?

Thanks

**Comment by nmcclatchey** [comment]
At the moment, `pool.schedule()` is roughly equivalent to `std::thread{}.detach();`. That is, it returns immediately without waiting for the task to complete. Because of this, it is currently expected that the user will provide their own mechanism for detecting when tasks complete.

For my own projects, I have been using a hand-rolled equivalent of [C++20's `std::latch`](https://en.cppreference.com/w/cpp/thread/latch) class, in the following manner:

```
void outer_thread (void)
{
  latch task_group(number_of_subtasks);
  for (int i = 0; i < number_of_subtasks; ++i)
    pool.schedule([&task_group](){ inner_thread(); task_group.count_down(); });
  while (task_group.try_wait() && early_task_available)
    early_task();
  task_group.wait();
}
```

I will clean this class up and upload it for this purpose.

If this is insufficient, or if you believe it to be inefficient, let me know and I will investigate alternate approaches.

**Comment by nmcclatchey** [comment]
A C++11-compatible version of `std::latch` has been provided. See abf67f1e60e65716e1e58bc4eb53ccd7590a5179 .

Note that the implementation is not perfect; the mechanisms that it uses for waiting are a `mutex` and `condition_variable`. This can be reduced to a lighter-weight variant if either of [`WaitOnAddress (Win32)`](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress) or [`futex` (Linux)](https://man7.org/linux/man-pages/man2/futex.2.html) is available, with additional programmer effort. Corresponding lightweight waiting features are made available in C++20, but so is a standard-library implementation of `latch`.

**Comment by tom-leys** [comment]
Thanks for taking the time to research my issue and write up a good solution. 

I can see how this works when the tasks know about how they are being coordinated outside the task. The solution I found between when I raised the issue and when you replied was to use std::packaged_task. I bet it has a much higher performance impact than your solution.

https://en.cppreference.com/w/cpp/thread/packaged_task

--- 
Offtopic: 

I hacked up [another thread-pool implementation](https://github.com/inkooboo/thread-pool-cpp/) so that I could "wait" in an inner task by calling a function in the thread-pool which is essentially cooperative multitasking. 

I was worried that any other form of waiting would eventually starve the pool of threads when all threads were waiting for other tasks to complete. 

My work often involves

Task A - Calls 10 instances of Task B
Task B - Calls 10+ inner tasks, waits for them to finish.

Task B must be able to wait "inside" a task but still make progress. 

---

Anyway, that's outside this issue. I suggest you make a couple of examples using std::latch and std::packaged_task and close this issue. Thanks so much for your help.

**Comment by nmcclatchey** [comment]
Example added; see commit 27b6ba936ea3024e6fa263e93de5736dcc674380.

"Waiting" is somewhat more tricky, as the state of the thread that invokes "wait" must be preserved. To accomplish this effect in pure standard C++11, a new thread would need to be spun off prior to each suspension, to ensure that the number of active threads does not decrease to zero. [C++20's coroutines](https://en.cppreference.com/w/cpp/language/coroutines) could potentially allow creation of a "pool::yield" function that schedules a continuation, though care would need to be taken to ensure that this continuation is executed in the same thread as the original (to avoid easily-made programming mistakes).
